ENH: integrate.cubature: array API standard support

ollybritton opened the pull request:
#### Reference issue
Builds on #21330 , towards gh-20930.

#### What does this implement/fix?
This adds support for the array API in `scipy.integrate.cubature` (currently in the open PR #21330). This means that functions returning arrays from any array API-compatible library can be used. The corresponding tests have been updated using the `@array_api_compatible` decorator.

#### Additional information
There are a few places where `np` is still used, such as in `integrate/_rules/_base.py::_subregion_coordinates`. Once CI shows whether the current changes work with other array APIs, I'll address these.

The first PR also introduced a (private) rules API, which allows different integration rules to be used with `cubature`. Rules need to know the correct array namespace to generate nodes and weights in, so these have been updated to accept an additional `xp` parameter on initialisation. This defaults to `np` if not set by the user. This is only a concern when using the rules API directly rather than using `cubature`, since `cubature` knows the array namespace being used by looking at the integration endpoints.

#### Acknowledgements
Thanks to @ev-br and @lucascolley for help with array API support.

ollybritton committed:
ENH: preliminary support for array api in cubature

ollybritton committed:
ENH: pass tests with array_api_strict

ollybritton committed:
BUG: don't use `arr.size` or `expand_dims`

ollybritton committed:
TST: add tests to array_api workflow

ollybritton committed:
BUG: remove unnecessary `asarray` in Genz-Malik

Removes an extra `xp.asarray` inside `xp.concat` which was causing
errors due to setting an array with a sequence.

ollybritton committed:
BUG: address array api compatibility problems

- Use `scipy._lib.array_api_compat.numpy` rather than `numpy` as a
  default when no array specified
- Remove use of in-place item assignment in `_cartesian_product`
- Remove use of `xp.newaxis` since Torch doesn't support this
- Switch to using positional arguments for `xp.linspace` since Torch
  uses different keywords
- Modify slow tests

lucascolley commented on scipy/integrate/_rules/_newton_cotes.py:
```diff
@@ -63,26 +63,31 @@ def __init__(self, npoints, open=False):
+    b = (1 - (-xp.ones(order+1))**(i + 1)) / (i + 1)
 
-    return np.linalg.solve(a, b)
+    return xp.linalg.solve(a, b)
```
`xp.linalg` is only an optional extension, so I think we want something like:

```python
if hasattr(xp, linalg):
    return xp.linalg.solve(a, b)
else:
    return xp.asarray(np.linalg.solve(np.asarray(a), np.asarray(b)))
```

lucascolley commented on scipy/integrate/_cubature.py:
```diff
@@ -19,23 +20,27 @@
-    error: np.ndarray
-    a: np.ndarray
-    b: np.ndarray
+    estimate: object
+    error: object
+    a: object
+    b: object
+    _xp: object
```
Maybe use this like in `_lib._array_api`:

```python
from types import ModuleType

if TYPE_CHECKING:
    Array = Any  # To be changed to a Protocol later (see array-api#589)
    ArrayLike = Array | npt.ArrayLike

estimate: Array
...
b: Array
_xp: ModuleType
```

ollybritton replied:
Thanks for taking a look! I've had to add an `else: Array = object` since otherwise there's errors about `Array` not being defined when defining the dataclass if `TYPE_CHECKING = False`.

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -24,18 +25,23 @@
 def basic_1d_integrand(x, n):
-    x_reshaped = x.reshape(-1, 1, 1)
-    n_reshaped = n.reshape(1, -1, 1)
+    xp = array_namespace(x, n)
```
would it be possible to add an `xp` argument to these functions so that we don't need to use `array_namespace` when we already have `xp` in the test functions?

lucascolley reviewed (commented):
skimmed the diff, nice stuff!

ollybritton committed:
BUG: better typing, np_compat, address CI failures

- Use same style of typing as `_array_api` for `Array`
- Use `np_compat` from `_lib.array_api` rather than
  `_lib.array_api_compat.numpy`
- Default to `np.linalg` when `xp.linalg` doesn't exist
- Add `xp` argument to test functions rather than calling
  `array_namespace` every time
- Skip testing `array_like` inputs when the backend is not numpy
- Change how `FixedRule`, `NestedRule` and `NestedFixedRule` determine
  the array namespace

ev-br commented:
Ran `$ python dev.py test -b cupy -t scipy/integrate/tests/test_cubature.py` locally. There are several failures, but nothing stands out too much from what is already seen in the CI. Which is a good sign: there are only a few small issues altogether.
 
Details are under the fold.

<details>
=================================================================== short test summary info ===================================================================
FAILED scipy/integrate/tests/test_cubature.py::TestCubature::test_pass_str[newton-cotes-cupy] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubature::test_pass_str[trapezoid-cupy] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubature::test_stops_after_max_subdivisions[cupy] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem0-cupy-gk15-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem0-cupy-gk21-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem1-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem2-cupy-gk15-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem2-cupy-gk21-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem6-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem8-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem9-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem11-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem12-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem14-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem15-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem18-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_scalar_output[problem19-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape0-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape1-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape2-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape3-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape4-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape5-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape6-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem3-cupy-gk15-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape7-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem3-cupy-gk15-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape8-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem0-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem1-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem1-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem1-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem2-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem3-cupy-genz-malik-1e-05-0.0001] - TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem4-cupy-gk15-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem4-cupy-gk21-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestCubatureProblems::test_array_output[shape9-problem4-cupy-genz-malik-1e-05-0.0001] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[NewtonCotesQuadrature-rule_args0-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[NewtonCotesQuadrature-rule_args1-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[NewtonCotesQuadrature-rule_args2-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[NewtonCotesQuadrature-rule_args3-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[NewtonCotesQuadrature-rule_args4-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[NewtonCotesQuadrature-rule_args5-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[GaussLegendreQuadrature-rule_args6-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[GaussLegendreQuadrature-rule_args7-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[GaussLegendreQuadrature-rule_args8-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[GaussKronrodQuadrature-rule_args9-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_simple[GaussKronrodQuadrature-rule_args10-cupy] - AttributeError: module 'cupy' has no attribute 'pow'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_error_from_difference[rule_pair0-rule_pair_args0-cupy] - AttributeError: module 'cupy' has no attribute 'concat'
FAILED scipy/integrate/tests/test_cubature.py::TestRulesQuadrature::test_base_1d_quadratures_error_from_difference[rule_pair1-rule_pair_args1-cupy] - AttributeError: module 'cupy' has no attribute 'concat'
========================================================= 124 failed, 117 passed, 5 skipped in 4.65s ==========================================================
(scipy-dev) ev-br@qgpu3:~/repos/scipy$ ~

</details>

ev-br commented:
Also non-array-api failures with numpy < 2 look real (np.atan, np.concat etc)

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -0,0 +1,856 @@
+
+    return (-1)**ndim * 1/xp.prod(alphas, axis=-1) \
+        * xp.prod(
+            xp.atan((a - betas)/alphas) - xp.atan((b - betas)/alphas),
```
While `atan` exists in the standard, for tests `xp` is the unwrapped namespace, so this doesn't work for np<2. There are (at least) two patterns you can use instead:

```python
xp_test = array_namespace(a) # np<2 needs atan
...
xp_test.atan(...)
```

or

```python
atan = array_namespace(a).atan # np<2 needs atan
# I think this works, but maybe instead you need
# atan = getattr(array_namespace(a), 'atan')
...
atan(...)
```

Your choice!

ollybritton committed:
BUG: use `np`-free versions of utility functions

- use a `_concat` helper function to pick whether to use `xp.concat` or
  `xp.concatenate`
- use `xp_test = array_namespace(a)` for using `atan` since `atan` isn't
  supported in np < 2.
- compute subregion coordinates and cartesian products without using
  `np`-specific code

ev-br commented:
Needs a rebase now. Either `git rebase upstream/main` and fix conflicts, or start from main and `git cherry-pick` new commits.

lucascolley commented on scipy/integrate/_cubature.py:
```diff
@@ -207,28 +218,34 @@ def cubature(f, a, b, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
     # API. An example of this can be found in the class scipy.integrate._rules.Rule.
 
-    max_subdivisions = np.inf if max_subdivisions is None else max_subdivisions
+    # If a and b are ordinary Python lists, default to NumPy
+    if isinstance(a, list) and isinstance(b, list):
+        a = np_compat.array(a)
+        b = np_compat.array(b)
+
+    xp = array_namespace(a, b)
```
This extra handling seems unnecessary - `array_namespace(a_list)` already returns `np_compat`.

lucascolley commented on scipy/integrate/_rules/_base.py:
```diff
@@ -447,50 +456,72 @@ def _subregion_coordinates(a, b):
     # shape (npoints,) rather than (npoints, 1)
     if orig_nodes.ndim == 1:
-        orig_nodes = orig_nodes[:, np.newaxis]
+        orig_nodes = orig_nodes[:, None]
```
you can also use `xp.newaxis`, but happy either way.

lucascolley reviewed (commented):
(haven't looked at the test changes yet)

ollybritton committed:
MAINT: delete unnecessary `isinstance`, `_concat`

- Removes an unnecessary `isinstance(a, list)` check since by default
  `array_namespace(a_list)` is `np_compat`
- Deletes the function `_concat` which was used to fall back to
  `np.concatenate` if `np.concat` wasn't available, and instead make
  sure `np_compat` is being used

ollybritton commented on scipy/integrate/_rules/_base.py:
```diff
@@ -447,50 +456,72 @@ def _subregion_coordinates(a, b):
+        return xp.asarray(np_compat.concatenate(
+            [np_compat.asarray(arr) for arr in arrays],
+            axis=axis,
+        ))
```
Thanks for taking a look, I've removed this now. The problem was that when each a `Rule` like `GaussKronrodQuadrature` is initialised, there's no array passed in the constructor that can be used to determine `xp`. Instead each `Rule` has an extra `xp` parameter in its `__init__`:

```python
def __init__(self, npoints, xp=None):
    # ...
    if xp is None:
        xp = np_compat

    self.xp = xp
```

Then it can return nodes and weights in the correct array namespace. But when the tests set NumPy as the backend, it was initialising each rule with `np` rather than `np_compat`. I've modified the constructors for each of the rules so that it will check if `xp` is NumPy and set it to `np_compat` instead. Hopefully this is the right approach?

lucascolley replied:
Sounds like a good approach yes, although why do we need to add `or is_numpy(xp)` to the condition? Either `xp=None`, or one should pass in `np_compat === array_namespace(an_np_array_or_scalar)`.

ollybritton replied:
If the rule is initialised like `Rule(..., xp=np)` for NumPy version < 2 (as it is in tests by the `@array_api_compatible` decorator) then `xp.concat` won't exist.

lucascolley replied:
Then you may also need similar shims for other libraries. I think you'll either want to:

- require that a compatible namespace is passed to `xp`, and do so in the tests
- acquire a compatible namespace in the rules via `xp = array_namespace(xp.empty(0))`

ollybritton replied:
Thanks! I'll go for option number 2, I was just about to update my comment asking if there was a way of determining the compatible namespace like this.

lucascolley commented on scipy/integrate/_rules/_gauss_kronrod.py:
```diff
@@ -83,7 +83,7 @@ def __init__(self, npoints, xp=None):
 
-        if xp is None:
+        if xp is None or is_numpy(xp):
             xp = np_compat
```
why do we need `is_numpy` here?

ollybritton committed:
BUG: make each rule use compatible array namespace

- Determine the compatible version of the array namespace in each rule

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -91,39 +94,40 @@ def genz_malik_1980_f_2(x, alphas, betas):
-    return (-1)**ndim * 1/np.prod(alphas, axis=-1) \
-        * np.prod(
-            np.arctan((a - betas)/alphas) - np.arctan((b - betas)/alphas),
+    return (-1)**ndim * 1/xp.prod(alphas, axis=-1) \
+        * xp.prod(
+            xp_test.atan((a - betas)/alphas) - xp_test.atan((b - betas)/alphas),
             axis=-1
         )
```
could you use parentheses instead of a backslash for line continuation here?

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -643,8 +658,8 @@ def test_scalar_output(self, problem, rule, rtol, atol):
-    def test_array_output(self, problem, rule, shape, rtol, atol):
-        np.random.seed(1)
+    def test_array_output(self, problem, rule, shape, rtol, atol, xp):
+        rng = np_compat.random.default_rng(1)
```
(I think you would be fine to just use `np` here, but this is fine too)

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -693,132 +712,137 @@ class TestRules:
-            NestedFixedRule(GaussKronrodQuadrature(21), GaussKronrodQuadrature(21)),
-        ),
-        (
-            # 1D problem, 2D rule
-            np.array([0]),
-            np.array([1]),
-            GenzMalikCubature(2),
```
was it deliberate to remove this case?

ollybritton replied:
No, thanks for spotting!

lucascolley reviewed (commented):
looks really good overall!

ollybritton committed:
TST: add back removed test, style updates

- Add back test to check if a 1D problem with a 2D rule raises an appropriate error
- Style changes related to line continuation

ollybritton committed:
DOC: add `xp` parameter to rule docstring

lucascolley reviewed (commented):
is this "ready for review" now?

lucascolley commented on scipy/integrate/_rules/_gauss_legendre.py:
```diff
@@ -32,14 +38,21 @@ class GaussLegendreQuadrature(FixedRule):
     @cached_property
     def nodes_and_weights(self):
-        return roots_legendre(self.npoints)
+        # TODO: current converting to/from numpy
+        nodes, weights = roots_legendre(self.npoints)
+        return self.xp.asarray(nodes), self.xp.asarray(weights)
```
is there test coverage for this? Typically, if we have to convert to NumPy and back, there are tests that need to be made cpu only.

lucascolley replied:
Ah so the "conversion to NumPy" is `self.npoints` giving us an `int`, right? So this is waiting for `roots_legendre` to get an `xp` parameter, or something like that? All good if so.

lucascolley reviewed (commented):
Looks great! Just one comment, will try running the tests on CUDA now.

lucascolley reviewed (approved):
`pixi run test-cuda -t scipy.integrate.tests.test_cubature` passes (after a while :sweat_smile: ) so all good from my side!

ev-br commented:
Free-threaded CI failure is unrelated, merging as approved. Thanks @ollybritton , @lucascolley for the  review
